### DISCUSSION TITLE
BoundsError during `eval_tree_array`

### DISCUSSION THREAD

**Discussion by DenisSvirin** [MilesCranmer/PySR#336] [Q&A] [answered]
I make new inputs `r = [6.332061275761631 5.051237472144821 2.7 4.2690748412273125 3.3068111527572905 5.4 4.676537180435969 6.037383539249433 5.727564927611035 3.818376618407357 1.9091883092036785]`.

and execute `_eval_tree_array`
like this `result, finished =  _eval_tree_array(tree, r, operators, (turbo ? Val(true) : Val(false))) `

its working for some time and then  error occurs. 
` BoundsError: attempt to access 1×11 Matrix{Float32} at index [2, 1:11]`

I understood that it comes from `deg0_eval` -> `return (cX[tree.feature, :], true)`

May be i should define `r` differently ? i don't understand where it finds other features

**Comment by MilesCranmer** [comment, marked as answer]
What is your `tree`? It looks like your expression has a feature=2 node, which is out-of-bounds for your input data. i.e., you would need to provide a 2D array with dimensions 2x11 instead.

**Reply by DenisSvirin** [reply]
But doesn't tree change?then it will work for different tree? if i understand correctly, feature is just index of variable from the input set and in my case it shouldn't be any. Why does it think that i have multiple variables? 
`r` is the same as `cX` in the function

**Reply by MilesCranmer** [reply]
When you call `eval_tree_array`, the second argument is the input data in the shape `(num_features, num_rows)`. So if your tree has any node with `tree.degree==0 && !tree.constant`, then `tree.feature` will be used to index the input data. So if you create a node with `tree.feature > num_features`, you will get some error like this. Does that make sense?

**Reply by DenisSvirin** [reply]
yes, but why does such tree appear? if `r` is just `cX` but with different values

**Reply by MilesCranmer** [reply]
Could you elaborate? Where are you seeing this behavior? I thought you were manually creating the tree here.

**Reply by DenisSvirin** [reply]
No, i just created `r` inside `eval_tree_array` and then passed it to the`_eval_tree_array`

**Reply by MilesCranmer** [reply]
But why create `r` as a 1x11 array instead of an array the same size as `cX`?

**Reply by DenisSvirin** [reply]
its necessary for the target function, as i need to evaluate functions not over `cX`, but `r` which is calculated from `cX`.`r` can change its length

**Reply by MilesCranmer** [reply]
`cX` has what shape?

**Reply by DenisSvirin** [reply]
1x 40

**Reply by MilesCranmer** [reply]
Hm, that’s a bug then. There’s no way for any tree to be created that indexes larger than first axis. Can you put all your code into a bug report?

**Reply by DenisSvirin** [reply]
ok
